cpp/ql/src/Security/CWE/CWE-704/WcharCharConversion.ql

```diff
@@ -13,23 +13,87 @@
  */
 
 import cpp
+import semmle.code.cpp.controlflow.Guards
 
 class WideCharPointerType extends PointerType {
   WideCharPointerType() { this.getBaseType() instanceof WideCharType }
 }
 
+/**
+ * Recurse through types to find any intermediate type or final type
+ * that suggests the type is unlikely to be a string.
+ * Specifically looking for any unsigned character, or datatype with name containing "byte"
+ * or datatype uint8_t.
+ */
+predicate hasIntermediateType(Type cur, Type targ) {
+  cur = targ
```

Comment:
Ideally this should have some restriction on the base case to limit the size of the predicate for performance reasons. For example, all places where we call this satisfies:
```ql
targ.(CharType).isUnsigned() or
targ.getName().toLowerCase().matches(["uint8_t", "%byte%"]) or
targ.getName() in [
    "LPCTSTR",
    "LPTSTR",
    "PCTSTR",
    "PTSTR",
    "TBYTE",
    "TCHAR"
  ]
```
This will dramatically lower the size of the predicate (and thus make it run much faster). However:
- There's a chance that the compiler already did this optimization - I didn't check!
- It's probably not worth it since there aren't _that_ many typedefs or derivations or a given type.

Reply:
Should the predicate be inlined you think? Or specify a bind order on the inputs?

Reply:
Sorry, I missed this comment. It's not possible to inline it (or give it any bindingsets) since it's recursive. So if we wanted to optimize this the only option we have is to constrain the base case (which is really the number one tip on how to optimize recursive predicates in QL).